### PR TITLE
Make InputParams (aka jobParams) functional

### PR DESCRIPTION
**Pull request by warren-jones**
New code in this branch enables passing provider-specific custom job parameters to an execution target from a Jupyter notebook or a Python host program.  The user may now specify "jobParams=" followed by one or more JSON {"key":"value",...} pairs when submitting a job.  This adds functionality to the work started in [PR 526](https://github.com/microsoft/iqsharp/pull/526)

### CODE REVIEW

File: src/AzureClient/Magic/ExecuteMagic.cs
Comment (cgranade) [inline src/AzureClient/Magic/ExecuteMagic.cs:55]: The type `ImmutableDictionary<string, string>` has no meaning in this context; readers will be working either in a Q# notebook or in a Q# + Python workflow, neither of which has an exact mapping for that C# type. In general, the types listed in this documentation string are really intended to be informal summaries for users as to what kind of input users can expect to pass.

File: src/AzureClient/Magic/SubmitMagic.cs
Comment (cgranade) [inline src/AzureClient/Magic/SubmitMagic.cs:54]: Same here.

**Review by cgranade — COMMENTED** [review]
Two initial thoughts:

- I don't see any change to unit tests, would you please add coverage for the new parameter?
- How will this be accessible from Q# + Python workflows? What changes to `qsharp/clients/iqsharp.py` need to be made for Python users?

File: src/AzureClient/Magic/ExecuteMagic.cs
Comment (warren-jones) [reply src/AzureClient/Magic/ExecuteMagic.cs:55]: Looking at unit tests now -- thought I'd get this initial code out there for comments while I worked on them.  I will address the Q# + Python workflow requirements next.

Regarding the references to ImmutableDictionary<string, string> in the documentation:  I was following the pattern used for the other parameters.  Should a simplified description of ImmutableDictionary<string, string> be substituted, or should that data type not be mentioned at all?

File: src/AzureClient/Magic/SubmitMagic.cs
Comment (warren-jones) [reply src/AzureClient/Magic/SubmitMagic.cs:54]: Please see reply to the ExecuteMagic.cs comment.

File: src/AzureClient/Magic/ExecuteMagic.cs
Comment (cgranade) [reply src/AzureClient/Magic/ExecuteMagic.cs:55]: > Looking at unit tests now -- thought I'd get this initial code out there for comments while I worked on them.

Makes sense, but in that case I'd suggest converting this to a draft PR so that it doesn't accidentally get merged before unit tests are added and the Q# + Python interop interface is modified for this new feature.

> Regarding the references to ImmutableDictionary<string, string> in the documentation: I was following the pattern used for the other parameters. Should a simplified description of ImmutableDictionary<string, string> be substituted, or should that data type not be mentioned at all?

The other entries use informal descriptions like "string" and "integer," since there's not any formal notion of types when calling magic commands from within a Q# notebook. In that sense, magic commands work a lot like command-line options (`--foo=bar`), such that simplified descriptions work well together with the more explicit examples later in the command documentation.

Comment (warren-jones) [reply src/AzureClient/Magic/ExecuteMagic.cs]: - Feature is supported in the Q# + Python interop interface
- Unit tests have been added (C# and Python)
- Documentation has been simplified and clarified

**Review by ricardo-espinoza — APPROVED** [review]
Looks good to me, but make sure you get approval from @anjbur before merging.

File: src/Tests/AzureClientMagicTests.cs
Comment (cgranade) [inline src/Tests/AzureClientMagicTests.cs:43]: This may be easier to read in terms of C#'s `@""` strings.

File: src/Tests/AzureClientMagicTests.cs
Comment (cgranade) [inline src/Tests/AzureClientMagicTests.cs:247]: This pattern happens enough that it may be worth adding a new `HasKey` extension method to the `Assert.That` object. Roughly,

```csharp
void HasKey<TKey, TValue>(this Assert assert, IDictionary<TKey, TValue> dict, TKey key, TValue expected)
{
     Assert.IsTrue(dict.TryGetValue(key, out TValue actual));
     Assert.AreEqual(actual, expected);
}
```

File: src/Python/qsharp-core/qsharp/tests/test_azure.py
Comment (cgranade) [inline src/Python/qsharp-core/qsharp/tests/test_azure.py:142]: Thanks for adding these tests; looks like the Python side picked up the new parameter correctly, then! Do we need to modify the docstrings accordingly?

File: src/AzureClient/AzureSubmissionContext.cs
Comment (warren-jones) [reply src/AzureClient/AzureSubmissionContext.cs]: Tags added as suggested.  Also applied these tags in lines 67 and 78 for consistency.

File: src/Jupyter/Extensions.cs
Comment (warren-jones) [reply src/Jupyter/Extensions.cs]: This line was eliminated by the following code re-write.

File: src/Jupyter/Extensions.cs
Comment (warren-jones) [reply src/Jupyter/Extensions.cs]: Eliminated the loop.  Simplified the code.

File: src/Tests/AzureClientMagicTests.cs
Comment (warren-jones) [reply src/Tests/AzureClientMagicTests.cs]: Good point, the `@` notation does look cleaner.  Also changed line 47 for consistency.

File: src/Tests/AzureClientMagicTests.cs
Comment (warren-jones) [reply src/Tests/AzureClientMagicTests.cs:247]: In my humble opinion, that extra layer of abstraction would make this code a little harder to read.  Granted, there is a bit of repetition, but not enough to significantly complicate maintenance.

File: src/Python/qsharp-core/qsharp/tests/test_azure.py
Comment (warren-jones) [reply src/Python/qsharp-core/qsharp/tests/test_azure.py:142]: The docstrings in [azure.py](https://github.com/microsoft/iqsharp/blob/main/src/Python/qsharp-core/qsharp/azure.py) refer to the magic docs, so we're covered.   Added jobParams to examples in the magic docs.

File: src/AzureClient/Magic/ExecuteMagic.cs
Comment (guenp) [inline src/AzureClient/Magic/ExecuteMagic.cs:100]: Should this be `--a=5 --b=10`

Comment (anjbur) [reply src/AzureClient/Magic/ExecuteMagic.cs:100]: The way inputs are currently parsed, adding `--` to the front would prevent the parameters from being recognized. It's a change we could theoretically make to the input parsing, but it'd be out of scope to this PR!

File: src/Tests/AzureClientMagicTests.cs
Comment (cgranade) [reply src/Tests/AzureClientMagicTests.cs:247]: Fair enough; I've found such abstractions to be highly useful in the code base so far, in part because it reduces friction to writing new tests that would otherwise introduce more duplication. This abstraction also reduces the probability of forgetting to check that the key exists when writing new tests.

That said, it is a minor comment, and something easily handled in code maintenance following this PR regardless.

File: src/AzureClient/Magic/ExecuteMagic.cs
Comment (cgranade) [reply src/AzureClient/Magic/ExecuteMagic.cs:100]: Ah, that's a good point; I hadn't noticed that the way its written, the new parameter strictly prevents adding `jobParams` as a Q# parameter name. That's probably minor, but we maybe should handle that better going forward. Since doing so would be a nontrivial breaking change, let's make sure to create an issue now and defer to that issue.

Comment (warren-jones) [reply src/AzureClient/Magic/ExecuteMagic.cs:100]: Resolving this conversation: Out of scope. The issue will be deferred.

File: src/Tests/AzureClientMagicTests.cs
Comment (ricardo-espinoza) [reply src/Tests/AzureClientMagicTests.cs:247]: Thanks a lot @cgranade ; for the purposes of this change let's leave this as is. I'll mark the comment as resolved. Going forward, it it makes sense to abstract this as part of feature work, let's do it. :)

**Review by anjbur — APPROVED** [review]
Everything looks good to me!

**Pull request merged**
